apps/indexer/src/api/controller/historical-balances.controller.ts

```diff
@@ -0,0 +1,209 @@
+import { OpenAPIHono as Hono, createRoute, z } from "@hono/zod-openapi";
+import { Address } from "viem";
+
+import { DaoIdEnum } from "@/lib/enums";
+import { caseInsensitiveEnum } from "../middlewares";
+import {
+  HistoricalBalancesService,
+  HistoricalBalance,
+  HistoricalBalancesRequest,
+} from "../services/historical-balances";
+
+export function historicalBalances(app: Hono) {
+  const service = new HistoricalBalancesService();
+
+  app.openapi(
+    createRoute({
+      method: "get",
+      operationId: "historicalBalances",
+      path: "/historical-balances/{daoId}",
+      summary: "Get historical token balances",
+      description:
+        "Fetch historical token balances for multiple addresses at a specific block number using multicall",
+      tags: ["historical-balances"],
+      request: {
+        params: z.object({
+          daoId: caseInsensitiveEnum(DaoIdEnum),
+        }),
+        query: z.object({
+          addresses: z.string().min(1, "At least one address is required"),
+          blockNumber: z.coerce.number().int().positive("Block number must be positive"),
```

Comment:
do we really need a positive verification on this?

Reply:
Why we would allow negative numbers?